homeassistant/components/binary_sensor/tod.py

```diff
@@ -0,0 +1,221 @@
+        else:
+            self._time_after = datetime.combine(
+                today.date(), self._after, tzinfo=self.hass.config.time_zone)
+
+        if is_sun_event(self._before):
+            before_event_date = get_astral_event_next(
+                self.hass, self._before, dt_util.as_utc(self._time_after))
+            if before_event_date:
+                self._time_before = dt_util.as_local(before_event_date)
+            else:
+                _LOGGER.error("Before event date for %s unknown", self._before)
+        else:
+            self._time_before = datetime.combine(
+                today.date(), self._before, tzinfo=self.hass.config.time_zone)
+            if self._time_after > self._time_before:
+                self._time_before += timedelta(days=1)
+
+        self._time_after += self._after_offset
+        self._time_before += self._before_offset
+
+    def _turn_to_next_day(self):
+        """Turn to to the next day."""
+        if is_sun_event(self._after):
+            self._time_after = dt_util.as_local(get_astral_event_next(
+                self.hass, self._after, dt_util.as_utc(
+                    self._time_after - self._after_offset)))
+            self._time_after += self._after_offset
+        else:
+            # offset is already there
+            self._time_after += timedelta(days=1)
+
+        if is_sun_event(self._before):
+            self._time_before = dt_util.as_local(get_astral_event_next(
+                self.hass, self._before, dt_util.as_utc(
+                    self._time_before - self._before_offset)))
+            self._time_before += self._before_offset
+        else:
+            self._time_before += timedelta(days=1)
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        self.point_in_time_listener(dt_util.now())
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def after(self):
+        """Return the timestamp for the begining of the period."""
+        return self._time_after
+
+    @property
+    def before(self):
+        """Return the timestamp for the end of the period."""
+        return self._time_before
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self.after < self.before:
+            if self.after <= self.current_datetime < self.before:
+                return STATE_ON
+        return STATE_OFF
+
+    @property
+    def current_datetime(self):
+        """Return local current datetime according to hass configuration."""
+        return dt_util.now(self.hass.config.time_zone)
+
+    @property
+    def next_change(self):
+        """Datetime when the next change to the state is."""
+        now = self.current_datetime
+        if now < self.after:
+            return self.after
+        if now < self.before:
+            return self.before
+
+        self._turn_to_next_day()
```

Comment:
A property access shouldn't modify anything IMO

Reply:
Will find another way. It is confusing.